feat: impl simple cli for wallet ops, sending

oetyng opened the pull request:
Resolves #128.

[feat: impl simple cli for wallet ops, sending](https://github.com/maidsafe/safe_network/commit/258624fdca16e4bafef77ede54b26ca3a4fbc925)

- Adds send operation.
- NB: Does not yet store the created dbcs, for giving them to the
recipients out of band.

joshuef commented:
superseded by #147

oetyng committed:
feat: impl simple cli for wallet ops, sending

- Adds send operation.
- NB: Does not yet store the created dbcs, for giving them to the
recipients out of band.

b-zee commented on safenode/src/client/wallet.rs:
```diff
@@ -30,9 +32,15 @@ impl<W: SendWallet> WalletClient<W> {
-        Ok(())
+    pub async fn send(&mut self, amount: Token, to: PublicAddress) -> Result<Vec<CreatedDbc>> {
+        let dbcs = self.wallet.send(vec![(amount, to)], &self.client).await?;
+        // self.wallet.store().await;
```
Any future ideas for this commented line?

oetyng replied:
aah... yeah... need to fix, let me see if I planned it for now or follow up..

oetyng replied:
Ah, it was a remainder. I had solved that already :)

oetyng replied:
(removing the line)

oetyng replied:
fixed

b-zee commented on safenode/src/bin/kadclient.rs:
```diff
@@ -138,6 +145,42 @@ async fn wallet(opt: &Opt) -> Result<()> {
+    if let Some(hex) = &opt.send_to {
+        let address = parse_public_address(hex)?;
+
+        if let Some(amount) = opt
```
Not sure if you're planning on erroring if the `send_amount` isn't given? I'm fine with this for now, but the user won't get to know if he specifies `send_to` but forgets the amount.

oetyng replied:
Good point, I can update that now already.

oetyng replied:
Will get some conflicts now with next PR but ye.

oetyng replied:
fixed

b-zee commented on safenode/src/bin/kadclient.rs:
```diff
@@ -138,6 +145,42 @@ async fn wallet(opt: &Opt) -> Result<()> {
+            .clone()
+            .map(|amount| parse_tokens_amount(&amount))
+        {
+            if amount.as_nano() > 0 {
```
Same here perpaps, show user an error or message if they specify `0`. I am also in favour of handling such cases with an early return to make sure we keep the code indentation low.

oetyng replied:
Ah all that is fixed in the next PR (with less indentation, I'm a stickler for that as well ;) )

oetyng replied:
Probably good to check the next one to see what stuff from here, is fixed there.

b-zee commented on safenode/src/bin/kadclient.rs:
```diff
@@ -149,6 +192,33 @@ async fn get_client_dir() -> Result<PathBuf> {
+        },
+    }
+
+    Token::from_nano(0)
```
Isn't it perhaps better to just straight error instead of defaulting to 0? I know the reason gets printed, but I'm not sure it's good to randomly default like this.

oetyng replied:
fixed

b-zee commented on safenode/src/protocol/wallet/keys.rs:
```diff
@@ -19,7 +19,15 @@ const MAIN_KEY_FILENAME: &str = "main_key";
 const PUBLIC_ADDRESS_FILENAME: &str = "public_address";
 
+/// Parse a public address from a hex-encoded string.
+#[allow(clippy::result_large_err)]
```
Is there a reason the `Error` grew so big?

oetyng replied:
It's something (from a while back) but didn't delve into it. Deferred.

oetyng committed:
refactor: error on cli invalid amount

- If sending tokens and an address has been specified, we will error if the
amount can't be parsed.
